File: src/shared/segmentation.ts
// ============================================================
// Pesto Captions — Segmentation Engine
// Pure function: no side effects, fully unit-testable.
// Converts word/phrase tokens into CaptionCues with casing
// and punctuation normalization applied.
// ============================================================

import type {
  WordToken,
  PhraseToken,
  CaptionCue,
  TextRun,
  SegmentationConfig,
  CasingMode,
  PunctuationConfig,
} from './types'

// ------------------------------------------------------------------
// Internal helpers
// ------------------------------------------------------------------

function applyCasing(text: string, mode: CasingMode): string {
  switch (mode) {
    case 'uppercase':
      return text.toUpperCase()
    case 'lowercase':
      return text.toLowerCase()
    case 'sentence': {
      const trimmed = text.trimStart()
      if (!trimmed) return text
      return text.slice(0, text.length - trimmed.length) + trimmed[0].toUpperCase() + trimmed.slice(1).toLowerCase()
    }
    default:
      return text
  }
}

function normalizePunctuation(text: string, cfg: PunctuationConfig): string {
  let t = text
  // Remove punctuation types that are disabled (set to false → strip from text)
  if (!cfg.comma) t = t.replace(/,/g, '')
  if (!cfg.period) t = t.replace(/\./g, '')
  if (!cfg.questionMark) t = t.replace(/\?/g, '')
  if (!cfg.exclamationMark) t = t.replace(/!/g, '')
  if (!cfg.quotes) t = t.replace(/["""'']/g, '')
  if (!cfg.dash) t = t.replace(/[-–—]/g, '')
  if (!cfg.semicolon) t = t.replace(/;/g, '')
  if (!cfg.colon) t = t.replace(/:/g, '')
  // Collapse double spaces produced by stripping
  t = t.replace(/  +/g, ' ').trim()
  return t
}

function parseEmphasisRuns(text: string): TextRun[] {
  // Parse **word** markup into runs
  const runs: TextRun[] = []
  const regex = /\*\*(.+?)\*\*/g
  let lastIndex = 0
  let match: RegExpExecArray | null

  while ((match = regex.exec(text)) !== null) {
    if (match.index > lastIndex) {
      runs.push({ text: text.slice(lastIndex, match.index), emphasis: false })
    }
    runs.push({ text: match[1], emphasis: true })
    lastIndex = regex.lastIndex
  }

  if (lastIndex < text.length) {
    runs.push({ text: text.slice(lastIndex), emphasis: false })
  }

  return runs.length > 0 ? runs : [{ text, emphasis: false }]
}

// ------------------------------------------------------------------
// Word-timed segmentation (Whisper path)
// ------------------------------------------------------------------

export function segmentFromWords(
  words: WordToken[],
  config: SegmentationConfig,
): CaptionCue[] {
  const cues: CaptionCue[] = []
  let group: WordToken[] = []
  let charCount = 0

  const flush = () => {
    if (group.length === 0) return

    let text = group.map(w => w.word).join(' ')
    text = normalizePunctuation(text, config.punctuation)
    text = applyCasing(text, config.casing)

    const startSec = group[0].start
    const endSec = group[group.length - 1].end
    const durationMs = (endSec - startSec) * 1000

    // Enforce minimum duration
    const effectiveEnd = durationMs < config.minDurationMs
      ? startSec + config.minDurationMs / 1000
      : endSec

    cues.push({
      cueIndex: cues.length + 1,
      startSec,
      endSec: effectiveEnd,
      runs: parseEmphasisRuns(text),
      timingEditable: true,
    })
    group = []
    charCount = 0
  }

  for (const word of words) {
    const wordText = word.word.trim()
    const wouldExceedChars = charCount + wordText.length + (group.length > 0 ? 1 : 0) > config.maxChars
    const wouldExceedWords = group.length >= config.maxWords

    if (group.length > 0 && (wouldExceedChars || wouldExceedWords)) {
      flush()
    }

    group.push(word)
    charCount += wordText.length + (group.length > 1 ? 1 : 0)
  }
  flush()

  // Gap filling: extend end of each cue to fill gap to next cue
  if (config.fillGapsMs > 0) {
    for (let i = 0; i < cues.length - 1; i++) {
      const gap = (cues[i + 1].startSec - cues[i].endSec) * 1000
      if (gap > 0 && gap <= config.fillGapsMs) {
        cues[i] = { ...cues[i], endSec: cues[i + 1].startSec }
      }
    }
  }

  // Enforce max duration (split long cues if needed — simple clamp)
  return cues.map(cue => {
    const dur = (cue.endSec - cue.startSec) * 1000
    if (dur > config.maxDurationMs) {
      return { ...cue, endSec: cue.startSec + config.maxDurationMs / 1000 }
    }
    return cue
  })
}

// ------------------------------------------------------------------
// Phrase-only segmentation (native Resolve path when no word-timing)
// ------------------------------------------------------------------

export function segmentFromPhrases(
  phrases: PhraseToken[],
  config: SegmentationConfig,
): CaptionCue[] {
  return phrases.map((phrase, i) => {
    let text = normalizePunctuation(phrase.text, config.punctuation)
    text = applyCasing(text, config.casing)

    const durationMs = (phrase.end - phrase.start) * 1000
    const effectiveEnd = durationMs < config.minDurationMs
      ? phrase.start + config.minDurationMs / 1000
      : phrase.end

    return {
      cueIndex: i + 1,
      startSec: phrase.start,
      endSec: effectiveEnd,
      runs: [{ text, emphasis: false }],  // No emphasis without word-timing
      timingEditable: false,
    }
  })
}

// ------------------------------------------------------------------
// Main entry point
// ------------------------------------------------------------------

export function segmentCues(
  input: { words: WordToken[] } | { phrases: PhraseToken[] },
  config: SegmentationConfig,
): CaptionCue[] {
  if ('words' in input) {
    return segmentFromWords(input.words, config)
  }
  return segmentFromPhrases(input.phrases, config)
}

// ------------------------------------------------------------------
// SRT/VTT parsing
// ------------------------------------------------------------------

export function parseSRT(content: string): PhraseToken[] {
  const blocks = content.trim().split(/\n\n+/)
  const phrases: PhraseToken[] = []

  for (const block of blocks) {
    const lines = block.split('\n')
    if (lines.length < 3) continue

    const timingLine = lines.find(l => l.includes('-->'))
    if (!timingLine) continue

    const [startStr, endStr] = timingLine.split('-->')
    const start = parseSRTTimestamp(startStr.trim())
    const end = parseSRTTimestamp(endStr.trim())
    const text = lines.slice(lines.indexOf(timingLine) + 1).join(' ').trim()

    if (!isNaN(start) && !isNaN(end) && text) {
      phrases.push({ text, start, end })
    }
  }

  return phrases
}

export function parseVTT(content: string): PhraseToken[] {
  const body = content.replace(/^WEBVTT.*\n/, '').trim()
  return parseSRT(body)
}

function parseSRTTimestamp(ts: string): number {
  // HH:MM:SS,mmm or HH:MM:SS.mmm
  const clean = ts.replace(',', '.')
  const parts = clean.split(':')
  if (parts.length !== 3) return NaN
  const [h, m, s] = parts.map(Number)
  return h * 3600 + m * 60 + s
}

// ------------------------------------------------------------------
// Utility: format seconds to SRT timestamp string
// ------------------------------------------------------------------

export function formatSRTTimestamp(sec: number): string {
  const h = Math.floor(sec / 3600)
  const m = Math.floor((sec % 3600) / 60)
  const s = Math.floor(sec % 60)
  const ms = Math.round((sec % 1) * 1000)
  return `${pad(h)}:${pad(m)}:${pad(s)},${padMs(ms)}`
}

function pad(n: number) { return String(n).padStart(2, '0') }
function padMs(n: number) { return String(n).padStart(3, '0') }

export function exportSRT(cues: CaptionCue[]): string {
  return cues.map(cue => {
    const text = cue.runs.map(r => r.text).join('')
    return `${cue.cueIndex}\n${formatSRTTimestamp(cue.startSec)} --> ${formatSRTTimestamp(cue.endSec)}\n${text}`
  }).join('\n\n')
}

File: pesto-qt/src/segmentation.py
"""
Pesto Captions — Segmentation Engine (Python)
AGPL-3.0

Pure functions: no side effects, fully testable.
"""

import re
from typing import Optional


# ── Types ──────────────────────────────────────────────────────────

def apply_casing(text: str, mode: str) -> str:
    if mode == "uppercase":
        return text.upper()
    if mode == "lowercase":
        return text.lower()
    if mode == "sentence":
        t = text.strip()
        if not t:
            return text
        return t[0].upper() + t[1:].lower()
    return text  # unchanged


def normalize_punctuation(text: str, cfg: dict) -> str:
    t = text
    if not cfg.get("comma", True):       t = t.replace(",", "")
    if not cfg.get("period", True):      t = t.replace(".", "")
    if not cfg.get("questionMark", True):t = t.replace("?", "")
    if not cfg.get("exclamationMark", True): t = t.replace("!", "")
    if not cfg.get("quotes", True):      t = re.sub(r'[""\'\'„"«»]', "", t)
    if not cfg.get("dash", True):        t = re.sub(r'[-–—]', "", t)
    if not cfg.get("semicolon", True):   t = t.replace(";", "")
    if not cfg.get("colon", True):       t = t.replace(":", "")
    return re.sub(r"  +", " ", t).strip()


def parse_emphasis_runs(text: str) -> list:
    """Parse **word** markup into TextRun list."""
    runs = []
    last = 0
    for m in re.finditer(r"\*\*(.+?)\*\*", text):
        if m.start() > last:
            runs.append({"text": text[last:m.start()], "emphasis": False})
        runs.append({"text": m.group(1), "emphasis": True})
        last = m.end()
    if last < len(text):
        runs.append({"text": text[last:], "emphasis": False})
    return runs if runs else [{"text": text, "emphasis": False}]


def segment_from_words(words: list, cfg: dict) -> list:
    """
    Segment word-timed tokens into CaptionCues.
    cfg keys: maxChars, maxWords, minDurationMs, maxDurationMs, fillGapsMs, casing, punctuation
    """
    cues = []
    group = []
    char_count = 0

    def flush():
        nonlocal group, char_count
        if not group:
            return
        text = " ".join(w["word"] for w in group)
        text = normalize_punctuation(text, cfg.get("punctuation", {}))
        text = apply_casing(text, cfg.get("casing", "unchanged"))

        start_sec = group[0]["start"]
        end_sec = group[-1]["end"]
        dur_ms = (end_sec - start_sec) * 1000
        effective_end = end_sec if dur_ms >= cfg.get("minDurationMs", 500) \
            else start_sec + cfg.get("minDurationMs", 500) / 1000

        cues.append({
            "cueIndex": len(cues) + 1,
            "startSec": start_sec,
            "endSec": effective_end,
            "runs": parse_emphasis_runs(text),
            "timingEditable": True,
        })
        group = []
        char_count = 0

    for word in words:
        w = word["word"].strip()
        extra = len(w) + (1 if group else 0)
        if group and (char_count + extra > cfg.get("maxChars", 42) or len(group) >= cfg.get("maxWords", 8)):
            flush()
        group.append(word)
        char_count += extra

    flush()

    # Fill gaps
    fill_ms = cfg.get("fillGapsMs", 0)
    if fill_ms > 0 and len(cues) > 1:
        for i in range(len(cues) - 1):
            gap_ms = (cues[i + 1]["startSec"] - cues[i]["endSec"]) * 1000
            if 0 < gap_ms <= fill_ms:
                cues[i]["endSec"] = cues[i + 1]["startSec"]

    # Clamp max duration
    max_ms = cfg.get("maxDurationMs", 5000)
    for cue in cues:
        if (cue["endSec"] - cue["startSec"]) * 1000 > max_ms:
            cue["endSec"] = cue["startSec"] + max_ms / 1000

    return cues


def segment_from_phrases(phrases: list, cfg: dict) -> list:
    """Segment phrase-level tokens (no word timing) into CaptionCues."""
    cues = []
    for i, phrase in enumerate(phrases):
        text = normalize_punctuation(phrase["text"], cfg.get("punctuation", {}))
        text = apply_casing(text, cfg.get("casing", "unchanged"))
        dur_ms = (phrase["end"] - phrase["start"]) * 1000
        effective_end = phrase["end"] if dur_ms >= cfg.get("minDurationMs", 500) \
            else phrase["start"] + cfg.get("minDurationMs", 500) / 1000
        cues.append({
            "cueIndex": i + 1,
            "startSec": phrase["start"],
            "endSec": effective_end,
            "runs": [{"text": text, "emphasis": False}],
            "timingEditable": False,
        })
    return cues


def segment_cues(transcribe_result: dict, cfg: dict) -> list:
    """Main entry: dispatch to word or phrase segmentation."""
    if transcribe_result.get("wordTimingAvailable") and transcribe_result.get("words"):
        return segment_from_words(transcribe_result["words"], cfg)
    return segment_from_phrases(transcribe_result.get("phrases", []), cfg)


# ── SRT/VTT parsing ────────────────────────────────────────────────

def _parse_ts(ts: str) -> float:
    ts = ts.strip().replace(",", ".")
    parts = ts.split(":")
    if len(parts) == 3:
        h, m, s = parts
        return int(h) * 3600 + int(m) * 60 + float(s)
    if len(parts) == 2:
        m, s = parts
        return int(m) * 60 + float(s)
    return float(parts[0])


def parse_srt(content: str) -> list:
    phrases = []
    for block in re.split(r"\n\n+", content.strip()):
        lines = block.strip().split("\n")
        timing_line = next((l for l in lines if "-->" in l), None)
        if not timing_line:
            continue
        parts = timing_line.split("-->")
        if len(parts) != 2:
            continue
        try:
            start = _parse_ts(parts[0])
            end = _parse_ts(parts[1].split()[0])  # strip extra metadata (VTT)
        except Exception:
            continue
        idx = lines.index(timing_line)
        text = " ".join(lines[idx + 1:]).strip()
        # Strip HTML tags that may appear in VTT
        text = re.sub(r"<[^>]+>", "", text).strip()
        if text:
            phrases.append({"text": text, "start": start, "end": end})
    return phrases


def parse_vtt(content: str) -> list:
    body = re.sub(r"^WEBVTT.*\n?", "", content, flags=re.MULTILINE).strip()
    return parse_srt(body)


# ── SRT export ─────────────────────────────────────────────────────

def format_ts(sec: float) -> str:
    h = int(sec // 3600)
    m = int((sec % 3600) // 60)
    s = int(sec % 60)
    ms = int(round((sec % 1) * 1000))
    return f"{h:02d}:{m:02d}:{s:02d},{ms:03d}"


def export_srt(cues: list) -> str:
    parts = []
    for cue in cues:
        text = "".join(r["text"] for r in cue["runs"])
        parts.append(f"{cue['cueIndex']}\n{format_ts(cue['startSec'])} --> {format_ts(cue['endSec'])}\n{text}")
    return "\n\n".join(parts)


# ── Default config ─────────────────────────────────────────────────

DEFAULT_PUNCTUATION = {
    "comma": True, "period": True, "questionMark": True,
    "exclamationMark": True, "quotes": True, "dash": True,
    "semicolon": True, "colon": True,
}

DEFAULT_SEGMENTATION = {
    "maxChars": 42,
    "maxWords": 8,
    "minDurationMs": 500,
    "maxDurationMs": 5000,
    "fillGapsMs": 0,
    "casing": "unchanged",
    "punctuation": DEFAULT_PUNCTUATION,
}
